src/snowflake/snowpark/_internal/utils.py

```diff
@@ -799,6 +807,155 @@ def warning(name: str, text: str, warning_times: int = 1) -> None:
+                if func.__qualname__.endswith(".__init__"):
+                    # Try to find a session argument.
+                    session_args = [
+                        arg
+                        for arg in args
+                        if isinstance(arg, snowflake.snowpark.session.Session)
+                    ]
+                    assert (
+                        len(session_args) != 0
+                    ), f"{func.__qualname__} must have at least one session arg."
+                    kwargs["_emit_ast"] = session_args[0].ast_enabled
+                else:
+                    kwargs["_emit_ast"] = args[0]._session.ast_enabled
+            elif isinstance(
+                args[0], snowflake.snowpark.dataframe_reader.DataFrameReader
+            ):
+                if func.__qualname__.endswith(".__init__"):
+                    assert isinstance(
+                        args[1], snowflake.snowpark.session.Session
+                    ), f"{func.__qualname__} second arg must be session."
+                    kwargs["_emit_ast"] = args[1].ast_enabled
+                else:
+                    kwargs["_emit_ast"] = args[0]._session.ast_enabled
+            elif isinstance(
+                args[0], snowflake.snowpark.dataframe_writer.DataFrameWriter
+            ):
+                if func.__qualname__.endswith(".__init__"):
+                    assert isinstance(
+                        args[1], snowflake.snowpark.DataFrame
+                    ), f"{func.__qualname__} second arg must be dataframe."
+                    kwargs["_emit_ast"] = args[1]._session.ast_enabled
+                else:
+                    kwargs["_emit_ast"] = args[0]._dataframe._session.ast_enabled
+            elif isinstance(
+                args[0],
+                (
+                    snowflake.snowpark.dataframe_stat_functions.DataFrameStatFunctions,
+                    snowflake.snowpark.dataframe_analytics_functions.DataFrameAnalyticsFunctions,
+                    snowflake.snowpark.dataframe_na_functions.DataFrameNaFunctions,
+                ),
+            ):
+                kwargs["_emit_ast"] = args[0]._dataframe._session.ast_enabled
+            elif hasattr(args[0], "_session") and args[0]._session is not None:
+                kwargs["_emit_ast"] = args[0]._session.ast_enabled
+            elif isinstance(args[0], snowflake.snowpark.session.Session):
+                kwargs["_emit_ast"] = args[0].ast_enabled
+            elif isinstance(
+                args[0],
+                snowflake.snowpark.relational_grouped_dataframe.RelationalGroupedDataFrame,
+            ):
+                kwargs["_emit_ast"] = args[0]._df._session.ast_enabled
+            else:
+                kwargs["_emit_ast"] = infer_ast_enabled_from_global_sessions(func)
+
+        # TODO: Could modify internal docstring to display that users should not modify the _emit_ast parameter.
```

Comment:
Do you mean customers should not modify? Other users like us or other developers can?